serde-generate/src/typescript.rs

```diff
@@ -0,0 +1,857 @@
+            F32 => "number".into(),
+            F64 => "number".into(),
+            Char => "string".into(),
+            Str => "string".into(),
+            Bytes => "Uint8Array".into(),
+
+            Option(format) => format!("Optional{}", Self::mangle_type(format)),
+            Seq(format) => format!("List{}", Self::mangle_type(format)),
+            Map { key, value } => {
+                format!("Map{}{}", Self::mangle_type(key), Self::mangle_type(value))
+            }
+            Tuple(formats) => format!(
+                "Tuple{}",
+                formats
+                    .iter()
+                    .map(Self::mangle_type)
+                    .collect::<Vec<_>>()
+                    .join("")
+            ),
+            TupleArray {
+                content,
+                size: _size,
+            } => format!("ListTuple{}", Self::mangle_type(content)),
+            Variable(_) => panic!("unexpected value"),
+        }
+    }
+
+    fn needs_helper(format: &Format) -> bool {
+        use Format::*;
+        match format {
+            Option(_) | Seq(_) | Map { .. } | Tuple(_) | TupleArray { .. } => true,
+            _ => false,
+        }
+    }
+
+    fn quote_serialize_value(&self, value: &str, format: &Format, use_this: bool) -> String {
+        use Format::*;
+        let this_str = if use_this { "this." } else { "" };
+
+        match format {
+            TypeName(_) => format!("{}{}.serialize(serializer);", this_str, value),
+            Unit => format!("serializer.serializeUnit({}{});", this_str, value),
+            Bool => format!("serializer.serializeBool({}{});", this_str, value),
+            I8 => format!("serializer.serializeI8({}{});", this_str, value),
+            I16 => format!("serializer.serializeI16({}{});", this_str, value),
+            I32 => format!("serializer.serializeI32({}{});", this_str, value),
+            I64 => format!("serializer.serializeI64({}{});", this_str, value),
+            I128 => format!("serializer.serializeI128({}{});", this_str, value),
+            U8 => format!("serializer.serializeU8({}{});", this_str, value),
+            U16 => format!("serializer.serializeU16({}{});", this_str, value),
+            U32 => format!("serializer.serializeU32({}{});", this_str, value),
+            U64 => format!("serializer.serializeU64({}{});", this_str, value),
+            U128 => format!("serializer.serializeU128({}{});", this_str, value),
+            F32 => format!("serializer.serializeF32({}{});", this_str, value),
+            F64 => format!("serializer.serializeF64({}{});", this_str, value),
+            Char => format!("serializer.serializeChar({}{});", this_str, value),
+            Str => format!("serializer.serializeStr({}{});", this_str, value),
+            Bytes => format!("serializer.serializeBytes({}{});", this_str, value),
+            _ => format!(
+                "Helpers.serialize{}({}{}, serializer);",
+                Self::mangle_type(format),
```

Comment:
For the same reason that `serializer.serializeU32` is not the same as `serializer.serializeU16`, we must have that `mangle_type(t1) != mangle_type(t2)` whenever `t1 != t2`. This property is already implemented in `common::mangle_type`. Please use this function and remove `Self::mangle_type`.

Reply:
There is no function overloading in typescript. this means there is no meaning to the type given to the seralizeX function. for that reason, I think there is no such constraint of mangle_type(t1) != mangle_type(t2) whenever t1 != t2 in regards to serializeX and serializeY.
anyway, I will change it to use only common::mangle_type

Reply:
What I meant is that if t1 != t2 then `Helpers.serialize{mangle_type(t1)}` and `Helpers.serialize{mangle_type(t2)}` must be distinct functions because the wire-format are (most likely) different.